File: VenueBitApp/VenueBitApp/Core/Networking/Models/Ticket.swift
import Foundation

struct Ticket: Codable, Identifiable {
    let id: String
    let eventId: String
    let eventTitle: String
    let eventDate: String
    let eventTime: String
    let venueName: String
    let section: String
    let row: String
    let seatNumber: Int
    let price: Double

    var seatDescription: String {
        "Section \(section), Row \(row), Seat \(seatNumber)"
    }

    var formattedDate: String {
        // Try parsing the date
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        guard let date = formatter.date(from: eventDate) else { return "\(eventDate) \(eventTime)" }

        let displayFormatter = DateFormatter()
        displayFormatter.dateFormat = "MMM d, yyyy"
        return "\(displayFormatter.string(from: date)) • \(eventTime)"
    }

    // Create ticket from order item and seat
    static func fromOrderItem(_ item: OrderItem, seat: OrderSeat) -> Ticket {
        Ticket(
            id: seat.id,
            eventId: item.eventId,
            eventTitle: item.eventTitle,
            eventDate: item.eventDate,
            eventTime: item.eventTime,
            venueName: item.venueName,
            section: seat.section,
            row: seat.row,
            seatNumber: seat.seatNumber,
            price: seat.price
        )
    }
}

struct OrderSeat: Codable, Identifiable {
    let id: String
    let section: String
    let row: String
    let seatNumber: Int
    let price: Double
}

struct OrderItem: Codable, Identifiable {
    let id: String
    let eventId: String
    let eventTitle: String
    let eventDate: String
    let eventTime: String
    let venueName: String
    let seats: [OrderSeat]
    let subtotal: Double
}

struct CartSeat: Codable, Identifiable {
    let id: String
    let eventId: String
    let section: String
    let row: String
    let seatNumber: Int
    let price: Double
    let status: String
}

enum SeatStatus: String, Codable {
    case available
    case reserved
    case sold
}

// API response types for seats
struct SeatsResponse: Codable {
    let success: Bool
    let data: [APISeat]
    let count: Int
}

struct APISeat: Codable, Identifiable {
    let id: String
    let eventId: String
    let section: String
    let row: String
    let seatNumber: Int
    let price: Double
    let status: String
}

File: VenueBitAndroid/app/src/main/java/com/venuebit/android/data/models/Ticket.kt
package com.venuebit.android.data.models

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass
import java.text.SimpleDateFormat
import java.util.Locale

/**
 * Represents a ticket for an event.
 */
@JsonClass(generateAdapter = true)
data class Ticket(
    val id: String,
    val eventId: String,
    val eventTitle: String,
    val eventDate: String,
    val eventTime: String,
    val venueName: String,
    val section: String,
    val row: String,
    val seatNumber: Int,
    val price: Double
) {
    /**
     * Returns a formatted seat description (e.g., "Section A, Row 1, Seat 5").
     */
    val seatDescription: String
        get() = "Section $section, Row $row, Seat $seatNumber"

    /**
     * Returns a formatted date string (e.g., "Aug 15, 2025 - 19:00").
     */
    val formattedDate: String
        get() {
            return try {
                val inputFormat = SimpleDateFormat("yyyy-MM-dd", Locale.US)
                val outputFormat = SimpleDateFormat("MMM d, yyyy", Locale.US)
                val date = inputFormat.parse(eventDate)
                date?.let { "${outputFormat.format(it)} - $eventTime" } ?: "$eventDate $eventTime"
            } catch (e: Exception) {
                "$eventDate $eventTime"
            }
        }

    companion object {
        /**
         * Creates a Ticket from an OrderItem and OrderSeat.
         */
        fun fromOrderItem(item: OrderItem, seat: OrderSeat): Ticket {
            return Ticket(
                id = seat.id,
                eventId = item.eventId,
                eventTitle = item.eventTitle,
                eventDate = item.eventDate,
                eventTime = item.eventTime,
                venueName = item.venueName,
                section = seat.section,
                row = seat.row,
                seatNumber = seat.seatNumber,
                price = seat.price
            )
        }
    }
}

/**
 * Represents a seat within an order.
 */
@JsonClass(generateAdapter = true)
data class OrderSeat(
    val id: String,
    val section: String,
    val row: String,
    val seatNumber: Int,
    val price: Double
)

/**
 * Represents an item within an order (contains multiple seats for an event).
 */
@JsonClass(generateAdapter = true)
data class OrderItem(
    val id: String,
    val eventId: String,
    val eventTitle: String,
    val eventDate: String,
    val eventTime: String,
    val venueName: String,
    val seats: List<OrderSeat>,
    val subtotal: Double
)
